pkg/token/filecache.go

```diff
@@ -0,0 +1,277 @@
+}
+
+var newFlock = func(filename string) filelock {
+	return flock.New(filename)
+}
+
+// cacheFile is a map of clusterID/roleARNs to cached tokens
+type cacheFile struct {
+	// a map of clusterIDs/profiles/roleARNs to cachedTokens
+	ClusterMap map[string]map[string]map[string]cacheToken `yaml:"clusters"`
+}
+
+// cacheToken is a single cached token entry, along with expiration time
+type cacheToken struct {
+	Token      credentials.Value
+	Expiration time.Time
+	// If set will be used by IsExpired to determine the current time.
+	// Defaults to time.Now if CurrentTime is not set.  Available for testing
+	// to be able to mock out the current time.
+	currentTime func() time.Time
+}
+
+// IsExpired determines if the cached token has expired
+func (t *cacheToken) IsExpired() bool {
+	curTime := t.currentTime
+	if curTime == nil {
+		curTime = time.Now
+	}
+	return t.Expiration.Before(curTime())
+}
+
+// readCacheWhileLocked reads the contents of the token cache and returns the
+// parsed yaml as a cacheFile object.  This method must be called while a shared
+// lock is held on the filename.
+func readCacheWhileLocked(filename string) (cache cacheFile, err error) {
+	cache = cacheFile{
+		map[string]map[string]map[string]cacheToken{},
+	}
+	data, err := f.ReadFile(filename)
+	if err != nil {
+		err = fmt.Errorf("unable to open file %s: %v", filename, err)
+		return
+	}
+
+	err = yaml.Unmarshal(data, &cache)
+	if err != nil {
+		err = fmt.Errorf("unable to parse file %s: %v", filename, err)
+	}
+	return
+}
+
+// writeCacheWhileLocked writes the contents of the token cache using the
+// yaml marshaled form of the passed cacheFile object.  This method must be
+// called while an exclusive lock is held on the filename.
+func writeCacheWhileLocked(filename string, cache cacheFile) error {
+	data, err := yaml.Marshal(cache)
+	if err == nil {
+		// write privately owned by the user
+		err = f.WriteFile(filename, data, 0600)
+	}
+	return err
+}
+
+// FileCacheProvider is a Provider implementation that wraps an underlying Provider
+// (contained in Credentials) and provides caching support for tokens for the
+// specified clusterID and roleARN
+type FileCacheProvider struct {
+	credentials                 *credentials.Credentials // the underlying implementation that has the *real* Provider
+	clusterID, profile, roleARN string                   // parameters used to create Provider
+	cachedToken                 cacheToken               // the cached token, if it exists
+}
+
+// NewFileCacheProvider creates a new Provider implementation that wraps a provided Credentials,
+// and works with an on disk cache to speed up token usage when the cached copy is not expired.
+// If there are any problems accessing or initializing the cache, an error will be returned, and
+// callers should just use the existing credentials provider.
+func NewFileCacheProvider(clusterID, profile, roleARN string, creds *credentials.Credentials) (FileCacheProvider, error) {
```

Comment:
does a lot of check to make sure things look good with the cache.  If any don't look right, it bails on the cache and reverts to existing behavior.